src/realm/db.cpp

```diff
@@ -1653,26 +1653,28 @@ class DB::AsyncCommitHelper {
     }
     ~AsyncCommitHelper()
     {
-        if (!m_terminated) {
-            {
-                std::unique_lock lg(m_mutex);
-                m_terminated = true;
-                m_changed.notify_one();
+        {
+            std::unique_lock lg(m_mutex);
+            if (!m_running) {
+                return;
             }
-            m_thread->join();
+            m_running = false;
+            m_changed.notify_one();
         }
+        m_thread->join();
     }
     void start_thread()
     {
-        if (m_terminated) {
+        {
             std::unique_lock lg(m_mutex);
-            if (m_terminated) {
-                m_thread = std::make_unique<std::thread>([this]() {
-                    main();
-                });
-                m_terminated = false;
+            if (m_running) {
+                return;
             }
+            m_running = true;
         }
+        m_thread = std::make_unique<std::thread>([this]() {
```

Comment:
since this is not under lock, in theory a concurrent call to the destructor could call join before the thread has been created.

Reply:
Creation of thread done while mutex is locked